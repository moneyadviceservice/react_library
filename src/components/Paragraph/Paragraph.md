### The `Paragraph` component should be used for all text on the page.
Many props are available to style paragraphs by default as shown in the examples below.

```jsx
import { Paragraph as P } from '@moneypensionservice/directories';

<>
  <P italic width='75%' margin={{bottom: '50px'}}>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam semper diam at erat pulvinar, at pulvinar felis blandit. Vestibulum volutpat tellus diam, consequat gravida libero rhoncus ut.</P>

  <P underline weight='700' textAlign='justify'>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam semper diam at erat pulvinar, at pulvinar felis blandit. Vestibulum volutpat tellus diam, consequat gravida libero rhoncus ut.</P>

  <P color='salmon' textAlign='right'>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam semper diam at erat pulvinar, at pulvinar felis blandit. Vestibulum volutpat tellus diam, consequat gravida libero rhoncus ut.</P>

  <P strike>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam semper diam at erat pulvinar, at pulvinar felis blandit. Vestibulum volutpat tellus diam, consequat gravida libero rhoncus ut.</P>
</>
```
