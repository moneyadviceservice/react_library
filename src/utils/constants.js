export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

