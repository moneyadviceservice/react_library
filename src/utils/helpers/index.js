export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
