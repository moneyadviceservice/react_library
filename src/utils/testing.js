import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
