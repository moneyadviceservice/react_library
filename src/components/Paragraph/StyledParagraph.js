import styled, { css } from 'styled-components'
import {
  genericStyles,
  breakpointStyle,
  responsiveProps,
} from '../../utils/helpers'
import { breakpoints } from '../../utils/constants'

const fontStyles = props => {
  const sizeStyles = props.theme.sizes.paragraph.sizing

  return Object.entries(sizeStyles).map(breakpoint => {
    const { size, height, marginTop, marginBottom } = breakpoint[1]

    const styles = css`
      ${!props.textSize && `font-size: ${size};`}
      ${!props.lineHeight && `line-height: ${height};`}
      ${!props.margin && `margin-top: ${marginTop};`}
      ${!props.margin && `margin-bottom: ${marginBottom};`}
    
      
    `
    // breakpoint styles
    return breakpointStyle(breakpoints[breakpoint[0]], styles)
  })
}

const StyledParagraph = styled.p`
  /** align-self, padding, margin, border */
  ${genericStyles}

  /** font styles */
  ${props => fontStyles(props)}
  ${props => props.italic && `font-style: italic;`}
  ${props => props.strike && `text-decoration: line-through;`}
  ${props => props.underline && `text-decoration: underline;`}

  /** conditional styles */
  ${props => props.weightProp && `font-weight: ${props.weightProp};`}
  color: ${props => props.colorProp || props.theme.colors.black};

  /** responsive props */
  ${props => props.textSize && responsiveProps('font-size', props.textSize)}
  
  ${props =>
    props.lineHeight && responsiveProps('line-height', props.lineHeight)}
  ${props => props.textAlign && responsiveProps('text-align', props.textAlign)}
  ${props => responsiveProps('width', props.widthProp)}

 
`

export { StyledParagraph }
