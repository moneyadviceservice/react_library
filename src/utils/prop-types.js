import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
