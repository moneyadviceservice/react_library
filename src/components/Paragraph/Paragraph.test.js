import React from 'react'
import { render } from '../../utils/testing'
import { Paragraph } from '../Paragraph'

describe('<Paragraph />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with children element', () => {
      const { container } = render(<Paragraph>example child</Paragraph>)
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with italic, width and margin props', () => {
      const { container } = render(
        <Paragraph italic width="75%" margin={{ bottom: '50px' }}>
          example child
        </Paragraph>
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with underline, weight and textAlign props', () => {
      const { container } = render(
        <Paragraph underline weight="700" textAlign="justify">
          example child
        </Paragraph>
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with color and strike props', () => {
      const { container } = render(
        <Paragraph color="salmon" strike>
          example child
        </Paragraph>
      )
      expect(container).toMatchSnapshot()
    })
  })
})
