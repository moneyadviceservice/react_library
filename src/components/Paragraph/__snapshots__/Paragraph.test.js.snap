// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paragraph /> Snapshots should match snapshot with children element 1`] = `
<div>
  <p
    class="StyledParagraph-sc-1toaej-0 gpltyO"
  >
    example child
  </p>
</div>
`;

exports[`<Paragraph /> Snapshots should match snapshot with color and strike props 1`] = `
<div>
  <p
    class="StyledParagraph-sc-1toaej-0 fKkfYQ"
  >
    example child
  </p>
</div>
`;

exports[`<Paragraph /> Snapshots should match snapshot with italic, width and margin props 1`] = `
<div>
  <p
    class="StyledParagraph-sc-1toaej-0 gZXJRU"
  >
    example child
  </p>
</div>
`;

exports[`<Paragraph /> Snapshots should match snapshot with underline, weight and textAlign props 1`] = `
<div>
  <p
    class="StyledParagraph-sc-1toaej-0 dtXQbJ"
  >
    example child
  </p>
</div>
`;
